swap/oracle.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import "time"
+
+const (
+	// maximum validity of the honey price
+	maxPriceAge       = 1 * time.Hour // TODO: this is currently an arbitrary value - currently irrelevant as prices are fixed
+	defaultHoneyPrice = uint64(1)     // default convertion of honey into output currency - currently ETH
+)
+
+// Currency represents a string identifier for currencies
+type Currency string
+
+// PriceOracle is the interface through which Oracles will deliver prices
+type PriceOracle interface {
+	GetPrice(honey int64) (int64, error)
+}
+
+// NewPriceOracle returns the actual oracle to be used for discovering the price
+// TODO: Add a config flag so that this can be configured via command line
+// For now it will return a default one
+func NewPriceOracle() PriceOracle {
+	cpo := &ConfigurablePriceOracle{
+		honeyPrice: defaultHoneyPrice,
+	}
+	cpo.refreshRate()
+	return cpo
+}
+
+// ConfigurablePriceOracle is a price oracle which can be customized.
+// It is the default price oracle used as a placeholder for this iteration of the implementation.
+// In production this should probably be some on-chain oracle called remotely
+type ConfigurablePriceOracle struct {
+	honeyPrice  uint64
+	lastUpdated time.Time
+}
+
+// GetPrice returns the actual price for honey
+func (cpo *ConfigurablePriceOracle) GetPrice(honey int64) (int64, error) {
+	if time.Now().After(cpo.lastUpdated.Add(maxPriceAge)) {
+		cpo.refreshRate()
```

Comment:
Why is the return value ignored here? Since `refreshState` also doesn't modify any state why call it at all? Wouldn't something like `cpo.honeyPrice = cpo.refreshRate()` make more sense?

Reply:
The `refreshRate` method is really a placeholder, as it is not doing anything really...it is only meant to give a suggestion on how the refreshing of exchanges via the (real) oracle *could* work in the future...in a real oracle refreshing the rate might be more complex and might take more time (Network request? File reload? I don't know!), so if the timeout has not happened yet, we would not call `refreshRate`.... 

When we have configurable prices, we could also change prices on the fly and then see if everything still works.

It is very possible this will also be requested to be removed on the merge to master (we usually request to remove code which is "for the future"), but for our experimental situation, I thought it might be interesting for now.

Reply:
`refreshRate` is not part of the Oracle interface, right? So it's an implementation detail of `ConfigurablePriceOracle` which for now is constant anyway. I would suggest to remove the `refreshRate` completely for now and use `cpo.honeyPrice` in `getPrice` directly.

Reply:
I finally agree this implementation mixes up concepts and is confusing at best. 

We agreed to have a fixed price "oracle" in the first iteration so let's do just that. I will rename `ConfigurablePriceOracle` to `FixedPriceOracle` and remove `refreshRate` completely.

There is now a new issue https://github.com/ethersphere/swarm/issues/1597 which will address having a configurable price oracle targeted mainly to dynamic testing in a deployed Swap-enabled testnet so that we can change prices dynamically and experiment.